File: aux_functions/admm.py
import numpy as np
import cvxpy as cp

def admm(X, Lt, gradient, Htp1, taut, dt, beta, verbose=False):
    S = len(taut)
    # variable
    L = cp.Variable(Lt.shape)
    N = Lt.shape[0]
    # constraints
    constraints = [cp.trace(L)== N, L.T==L, L@np.ones((N,1))==np.zeros((N,1))]
    for i in range(N-1):
        constraints += [L[i][i+1:]<=0]
    # objective
    obj = cp.Minimize(cp.trace(gradient.T@(L-Lt)) \
                      + dt/2*(cp.norm(L-Lt, 'fro')**2) + beta*(cp.norm(L, 'fro')**2))
    # solve problem
    prob = cp.Problem(obj, constraints)
    prob.solve(verbose=verbose, solver=cp.SCS, scale = 1000, use_indirect = False)
    if L.value is None:
        prob.solve(verbose=verbose, solver=cp.MOSEK)
    return L.value

L = admm(X, Lt, gradient, Htp1, taut, dt, beta)

File: aux_functions/admm.m
% trying to imitate admm.py with matlab

function L = admm(L_old, grad_L, lip_d, params)
    cvx_begin quiet
        variable L_new(params.n, params.n);
        minimize(trace(grad_L'*(L_new - L_old)) +...
            (lip_d/2)*pow_pos(norm(L_new-L_old,"fro"),2) +...
            params.beta*pow_pos(norm(L_new,"fro"),2));
        subject to;
            trace(L_new) == params.n;
            L_new == L_new';
            for i=1:params.n
                for j=1.params.n
                    L_new(i,j) <= 0;
                end
            end
            L_new*ones(params.n,1) == zeros(params.n,1);
    cvx_end
end